README.md
# MOMOpack for R

# News

## MOMO 2020.9.3

Spring weeks of 2020, COVID-19, permanently excluded from baseline estimation

## MOMO 2020.4.17

Ydrop and Wdrop as parameters
glmnet removed from list of dependencies

## MOMO 2018.8.24

* There are now more options for extracting data from inside your MoMo run. Previously only `MOMO::dataExport$toSave` was available, however, now there is also `MOMO::dataExport$aggr`, `MOMO::dataExport$aggr_fullDelay`, and `MOMO::dataExport$aggr_delay`

## MOMO 2018.8.23

* The new delay correction that Jens developed in December 2017 is now available. This can be chosen by `delayVersion="2017-12"`, or the original delay correction can be chosen by `delayVersion="original"` (default).
    
## Initial: MOMO 2017.12.14

* This is a full port of the Stata A-MOMO package to R. The code used to analyze mortality data for the [EuroMOMO project](www.euromomo.eu).


# How to install

Please use the [drat](https://github.com/eddelbuettel/drat) repository to install this package from, as the drat repository only contains package versions that have passed [testing](http://travis-ci.org/euromomonetwork/MOMO).

```
install.packages("MOMO", repos = c("https://euromomonetwork.github.io/drat", "https://cran.rstudio.com"))
```

# Basic example

```
SetOpts(
  DoA=as.Date("2013-12-31"),
  DoPR=as.Date("2008-1-1"),
  WStart=1,
  WEnd=52,
  country = "Denmark",
  source = "SSI",
  MFILE = "DoD_DoR.txt",
  HFILE = "holidays.txt",
  INPUTDIR = system.file("testdata", package="MOMO"),
  WDIR = tempdir(),
  back = 3,
  WWW = 290,
  Ysum = 2013,
  Wsum = 40,
  USEglm2 = TRUE,
  useAUTOMN = TRUE,
  datesISO = FALSE,
  plotGraphs = TRUE,
  delayVersion = "2017-12",
  MOMOgroups = list(`0to14` = "0 <= age & age <= 14",
                    `15to44` = "15 <= age & age <= 44",
                    `45to64` = "45 <= age & age <= 64",
                    `65P` = "age >= 65 & !is.na(age)",
                    `65to74` = "65 <= age & age <= 74",
                    `75to84` = "75 <= age & age <= 84",
                    `85P` = "age >= 85 & !is.na(age)",
                    `Total` = "age >= 0 | is.na(age)"
  ),
  MOMOmodels = c(`0to14` = "LINE",
                 `15to44` = "LINE_SIN",
                 `45to64` = "LINE_SIN",
                 `65P` = "LINE_SIN",
                 `65to74` = "LINE_SIN",
                 `75to84` = "LINE_SIN",
                 `85P` = "LINE_SIN",
                 `Total` = "LINE_SIN"
  ),
  Ydrop = 9999,
  Wdrop = 99,
  verbose = TRUE
)

RunMOMO()
```

# Another example using MDATA and HDATA instead of MFILE and HFILE

```
MDATA <- as.data.frame(data.table::fread(file.path(system.file("testdata",package="MOMO"),"DoD_DoR.txt")))
HDATA <- as.data.frame(data.table::fread(file.path(system.file("testdata",package="MOMO"),"holidays.txt")))

SetOpts(
  DoA=as.Date("2013-12-31"),
  DoPR=as.Date("2008-1-1"),
  WStart=1,
  WEnd=52,
  country = "Denmark",
  source = "SSI",
  MDATA = MDATA,
  HDATA = HDATA,
  WDIR = tempdir(),
  back = 3,
  WWW = 290,
  Ysum = 2013,
  Wsum = 40,
  USEglm2 = TRUE,
  useAUTOMN = TRUE,
  datesISO = FALSE,
  plotGraphs = TRUE,
  delayVersion = "2017-12",
  MOMOgroups = list(`0to14` = "0 <= age & age <= 14",
                    `15to44` = "15 <= age & age <= 44",
                    `45to64` = "45 <= age & age <= 64",
                    `65P` = "age >= 65 & !is.na(age)",
                    `65to74` = "65 <= age & age <= 74",
                    `75to84` = "75 <= age & age <= 84",
                    `85P` = "age >= 85 & !is.na(age)",
                    `Total` = "age >= 0 | is.na(age)"
  ),
  MOMOmodels = c(`0to14` = "LINE",
                 `15to44` = "LINE_SIN",
                 `45to64` = "LINE_SIN",
                 `65P` = "LINE_SIN",
                 `65to74` = "LINE_SIN",
                 `75to84` = "LINE_SIN",
                 `85P` = "LINE_SIN",
                 `Total` = "LINE_SIN"
  ),
  Ydrop = 9999,
  Wdrop = 99,
  verbose = TRUE
)

RunMOMO()
```

# Delay corrections available

| Delay correction | Description |
| --- | --- |
| original | ??? |
| sincos | original+sin52+cos52 |
| richard | Interacts WR with categorical YoDi, includes closed, sin52, and cos52nb2 ~ WRXX*as.factor(YoDi) + closed + sin52 + cos52 |

## original

???

## sincos

???

## richard

`nb2 ~ WRXX*as.factor(YoDi) + closed + sin52 + cos52`



NEWS.md
## MOMO 2020.9.3

Spring weeks of 2020, COVID-19, permanently excluded from baseline estimation

## MOMO 2020.4.27

Added the input variables Ydrop and Wdrop

## MOMO 2020.4.17

glmnet removed from list of dependencies

## MOMO 2018.8.24

* There are now more options for extracting data from inside your MoMo run. Previously only `MOMO::dataExport$toSave` was available, however, now there is also `MOMO::dataExport$aggr`, `MOMO::dataExport$aggr_fullDelay`, and `MOMO::dataExport$aggr_delay`

## MOMO 2018.8.23

* The new delay correction that Jens developed in December 2017 is now available. This can be chosen by `delayVersion="2017-12"`, or the original delay correction can be chosen by `delayVersion="original"` (default).
    
## MOMO 2017.12.14

* This is a full port of A-MOMOpack (the code used to analyze mortality data for the [EuroMOMO project](www.euromomo.eu)) from Stata to R.


R/Benchmark.R
#' Benchmark different delay corrections.
#' @param tag Tag for file name
#' @param wdir workingdir
#' @import data.table
#' @importFrom lubridate today
#' @export BenchmarkDelay
BenchmarkDelay <- function(tag,wdir=NULL){
  if(!opts$setByUser){
    stop("You have not set the options in function momo::SetOpts. You need to do this before using momo::RunMoMo")
  }

  # remove some of the more pointless functions for a benchmark
  if(is.null(wdir)) wdir <- opts$WDIR
  opts$WDIR <- tempdir()
  opts$verbose <- F

  RunMoMo()

  trueRes <- data.table::rbindlist(dataExport$toSave)
  predNames <- names(trueRes)[stringr::str_detect(names(trueRes),"^predzscore")]

  trueRes <- trueRes[,c("GROUP","YoDi","WoDi","excess","zscore","Pnb","UPIb2",predNames),with=F]

  res <- dataExport$aggr_fullDelay
  res <- merge(res,trueRes,by=c("GROUP","YoDi","WoDi"))
  setDT(res)

  opts$WDIR <- wdir

  modellingWeeks <- momoAttr$PRWEEK:momoAttr$WEEK2

  x <- res[wk %in% modellingWeeks]
  ra <- dataExport$raw

  rmarkdown::render(input = file.path(system.file("ext",package="MOMO"),"BenchmarkDelay.Rmd"),
                    output_file = sprintf("BenchmarkDelay_%s_%s_%s.pdf",opts$country,tag,lubridate::today()),
                    output_dir = wdir, output_format = rmarkdown::pdf_document(toc=TRUE),
                    params = list(x=x, ra=ra))

  return(x)
}

#' @import ggplot2
#' @importFrom readxl read_excel
BenchmarkNorway <- function(wdir=system.file("benchmarkdelay",package="MOMO")){
  #devtools::load_all()
  #library(data.table)
  #library(ggplot2)

  masterData <- fread("/docs/dashboards/data_raw/normomo/FHIDOD2_20180814.txt")
  masterData[,DoD:=as.Date(as.character(DODS_DATO),format="%Y%m%d")]
  masterData[,DoR:=as.Date(as.character(ENDR_DATO),format="%Y%m%d")]
  masterData[,DoB:=as.Date(as.character(FDATO_YYYYMMDD),format="%Y%m%d")]
  masterData[,age:=floor(as.numeric(difftime(DoD,DoB,units="days"))/365.25)]
  masterData[is.na(DoR),DoR:=DoD+1]
  masterData[DoR>="2015-09-03",DoR:=DoR+1]

  masterData[,ageCat:=cut(age,c(0,4,14,64,200),include.lowest = TRUE)]
  masterData[,deathWeek:=WeekN(masterData$DoD)]
  masterData[,deathYear:=YearN(masterData$DoD)]
  MDATA <- as.data.frame(masterData[!is.na(age),c("DoD","DoR","age")])

  HDATA <- data.frame(readxl::read_excel(system.file("extdata", "bank_holidays.xlsx", package = "normomo"))[,c("date", "closed")])
  HDATA$date <- as.Date(HDATA$date)

  for(dv in c("original","sincos","richard")){
  #for(dv in c("dummyyear")){
    SetOpts(
      DoA=as.Date("2018-08-14"),
      DoPR=as.Date("2012-01-01"),
      WStart=1,
      WEnd=52,
      country = "Norway",
      source = "FHI",
      MDATA = MDATA,
      HDATA = HDATA,
      INPUTDIR = tempdir(),
      WDIR = tempdir(),
      back = 7,
      WWW = 290,
      Ysum = 2018,
      Wsum = 40,
      USEglm2 = TRUE,
      useAUTOMN = TRUE,
      datesISO = FALSE,
      plotGraphs = FALSE,
      delayVersion = dv,
      MOMOgroups = list("Total"="age >= 0 | is.na(age)"),
      MOMOmodels = list("Total"="LINE_SIN"))
    #delayFunction = MOMO::delayMOMO_original,
    x <- BenchmarkDelay(tag=dv,wdir=wdir)
  }
}

BenchmarkDenmark <- function(wdir=system.file("benchmarkdelay",package="MOMO")){
  #devtools::load_all()
  #library(data.table)
  #library(ggplot2)

  MDATA <- as.data.frame(data.table::fread(file.path(system.file("testdata",package="MOMO"),"DoD_DoR.txt")))
  HDATA <- as.data.frame(data.table::fread(file.path(system.file("testdata",package="MOMO"),"holidays.txt")))

  for(dv in c("original","richard","richard1","richard2")){
    SetOpts(
      DoA=as.Date("2013-12-31"),
      DoPR=as.Date("2008-1-1"),
      WStart=1,
      WEnd=52,
      country = "Denmark",
      source = "SSI",
      MDATA = MDATA,
      HDATA = HDATA,
      INPUTDIR = tempdir(),
      WDIR = wdir,
      back = 7,
      WWW = 290,
      Ysum = 2018,
      Wsum = 40,
      USEglm2 = TRUE,
      useAUTOMN = TRUE,
      datesISO = FALSE,
      plotGraphs = FALSE,
      delayVersion = dv,
      MOMOgroups = list("Total"="age >= 0 | is.na(age)"),
      MOMOmodels = list("Total"="LINE_SIN"))

    x <- BenchmarkDelay(tag=dv)
  }
}

